### PR TITLE
[RLlib] Remove Unnecessary List Conversion of Complex Observations in SAC Model

### LINKED ISSUE #24105: [RLlib] Continuous SAC Messes up Order in Complex Dict-Observations
### What happened + What you expected to happen

I am using CQL with custom offline data containing complex dict observations. Note that CQL does not apply any preprocessors for complex spaces but uses the data as provided. With this setup, I ran into an error which is caused by the underlying SAC model.

SAC models may mix up the order of sub-observations provided as a complex Dict space before passing it to the Q-Network. This happens while combining the observations with continuous actions.

When building the Q-Networks for continuous actions, the values of Dict observation spaces are packed into a list using `list(orig_space.spaces.values())` (cf. [line 194, SACTorchModel](https://github.com/ray-project/ray/blob/dfb9689701361cfd18f383e0a3edeed6baf81abb/rllib/agents/sac/sac_torch_model.py#L194)). In Dict spaces, the values are always sorted lexicographically by keys.
This is not necessarily the case for the actual observations which might be `dict`s with arbitrarily ordered values. Therefore, when putting the values into a list via `list(model_out.values())` (cf. [line 257, SACTorchModel](https://github.com/ray-project/ray/blob/dfb9689701361cfd18f383e0a3edeed6baf81abb/rllib/agents/sac/sac_torch_model.py#L257)), the order of values can differ from the order of spaces.
In the subsequent pass through the (complex) Q-Network, the sub-observations are then passed to the wrong network inputs defined by the space.

### Versions / Dependencies

Ray nightly

### Reproduction script

The bug can also be reproduced with SAC and disabled preprocessor API:

```python
import gym

from ray.rllib.agents.sac import SACTrainer
from ray.tune.registry import register_env


DICT_SPACE = gym.spaces.Dict(
    {
        "cam": gym.spaces.Box(low=0, high=1, shape=(42, 42, 3)),
        "progress": gym.spaces.Box(low=0, high=100, shape=(3,)),
    }
)

# Dict space .sample() returns an ordered dict.
# Make sure the keys in samples are ordered differently to cause the bug.
DICT_SAMPLES = [{k: v for k, v in reversed(DICT_SPACE.sample().items())} for _ in range(10)]


class NestedDictEnv(gym.Env):
    def __init__(self):
        self.action_space = gym.spaces.Box(low=42, high=56, shape=(2,))
        self.observation_space = DICT_SPACE
        self._spec = gym.envs.registration.EnvSpec("NestedDictEnv-v0")
        self.steps = 0

    def reset(self):
        self.steps = 0
        return DICT_SAMPLES[0]

    def step(self, action):
        self.steps += 1
        return DICT_SAMPLES[self.steps], 1, self.steps >= 5, {}

# Build SAC trainer and apply to NestedDictEnv
register_env("nested", lambda _: NestedDictEnv())
sac = SACTrainer(
    env="nested",
    config={
        "num_workers": 0,
        "rollout_fragment_length": 5,
        "train_batch_size": 5,
        "replay_buffer_config": {
            "capacity": 10,
        },
        "learning_starts": 1,
        "framework": "torch",
        "_disable_preprocessor_api": True,
    },
)
sac.train()
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by fwitter**
## Why are these changes needed?

SAC models pack the values of complex observations into a list before passing them to the Q-Networks. This step may cause a bug for `dict` observations where the order of the keys is mixed up.
Moreover, this list conversion is not necessary. In case of continuous actions which need to be added to the input of the Q-Networks, we can also add another hierarchical layer on top of the complex observations to attach the action.

## Related issue number

Fixes #24105

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/sac/sac_tf_model.py [hunk 1]
```diff
@@ -180,13 +179,7 @@ def build_q_model(self, obs_space, action_space, num_outputs, q_model_config, na
...
                 self.concat_obs_and_actions = True
             else:
-                if isinstance(orig_space, gym.spaces.Tuple):
-                    spaces = list(orig_space.spaces)
-                elif isinstance(orig_space, gym.spaces.Dict):
-                    spaces = list(orig_space.spaces.values())
-                else:
-                    spaces = [obs_space]
-                input_space = gym.spaces.Tuple(spaces + [action_space])
+                input_space = gym.spaces.Tuple([orig_space, action_space])
```
Comment (sven1977) [inline rllib/agents/sac/sac_tf_model.py:182]: Nice!

File: rllib/agents/sac/sac_tf_model.py [hunk 2]
```diff
@@ -255,7 +246,7 @@ def _get_q_value(self, model_out, actions, net):
             else:
                 # TODO(junogng) : SampleBatch doesn't support list columns yet.
                 #     Use ModelInputDict.
```
Comment (sven1977) [inline rllib/agents/sac/sac_tf_model.py:248]:
```suggestion
```

File: rllib/agents/sac/sac_torch_model.py [hunk 3]
```diff
@@ -263,7 +254,7 @@ def _get_q_value(self, model_out, actions, net):
             else:
                 # TODO(junogng) : SampleBatch doesn't support list columns yet.
                 #     Use ModelInputDict.
```
Comment (sven1977) [inline rllib/agents/sac/sac_torch_model.py:256]:
```suggestion
```

**Review by sven1977 — APPROVED** [review]
A really nice, elegant fix! Thanks so much @fwitter. Makes the SAC code so much more readable. :)

**Comment by gjoliver** [comment]
yeah, I have to come in and manually +1 this. thanks for the awesome PR.

**Pull request merged**